Product Gallery > Pager block: Fix block alignment

thealexandrelara opened the pull request:
<!-- Please do not remove any information from this pull request. Instead, add N/A or leave blank if not applicable -->

## What
This PR fixes a bug that was causing the Pager block to appear misaligned with the Large Image block.

Fixes #10781

## Why
As defined in the design for the block, the Pager block must appear below the Large Image block and centered with it.

<!-- Describe the reason for your changes. This will help the reviewer and future readers get additional context -->

## Testing Instructions

<!-- Write these steps from the perspective of a "user" (merchant) familiar with WooCommerce. No need to spell out the steps for common setup scenarios (eg. "Create a product"), but do be specific about the thing being tested. Include screenshots demonstrating expectations where that will be helpful. -->

_Please consider any edge cases this change may have, and also other areas of the product this may impact._

1. From your WordPress dashboard, go to Appearance > Themes. Make sure you have a block-based theme installed and activated. If not, you can install one from the Add New option. Block-based themes include "Twenty-twenty Two," "Twenty-twenty Three," etc.
2. On the left-hand side menu, click on Appearance > Editor > Templates
3. Find and select the 'Single Product' template from the list.
4. When the Classic Product Template renders, click on Transform into Blocks. This will transform the Classic template in a block template if you haven't done it before.
5. Inside the Page editor, click on the '+' button, usually found at the top left of the editing space or within the content area itself, to add a new block.
6. In the block library that pops up, search for the 'Product Gallery' block. Click on it to add the block to the template.
7. Check that the Pager block is displayed on the Editor.
8. Select the Product Gallery block and change the Align mode on the toolbar to "Full width"
![CleanShot 2023-08-30 at 14 23 39@2x](https://github.com/woocommerce/woocommerce-blocks/assets/20469356/a9b22546-714d-4e9d-9b76-f32ea4687fd8)
9. Make sure the block is correctly centered with the Large Image block.
10. Save the page and visit the product page, make sure the block is correctly being displayed and centered with the Large Image block.

* [x] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested (ie: it makes changes to tests, coding standards, docblocks, etc.). -->
* [ ] Should be tested by the development team exclusively <!-- Check this box if this PR should be tested by the development team exclusively (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

## Screenshots or screencast

<!-- Any screenshots of UI changes will be helpful to include here. Leave blank if not applicable. -->

| Before | After |
| ------ | ----- |
| <img width="1595" alt="CleanShot 2023-08-30 at 17 04 32@2x" src="https://github.com/woocommerce/woocommerce-blocks/assets/20469356/bed73658-7402-4e56-b88e-4f4926fa6456"> | <img width="1096" alt="CleanShot 2023-08-30 at 16 42 43@2x" src="https://github.com/woocommerce/woocommerce-blocks/assets/20469356/d776f315-4d40-452d-86f8-d253fa3872c5"> |

## WooCommerce Visibility

<!-- Check this documentation link (../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WooCommerce core, part of the feature plugin, or experimental. -->
Required:

* [ ] WooCommerce Core
* [ ] Feature plugin
* [x] Experimental
* [ ] N/A

thealexandrelara committed:
Fix block alignment with Large Image block

thealexandrelara committed:
Prevent error when layout is not present in the attributes object

danieldudzic reviewed (changes requested):
In the editor it seems that the vertical alignment settings are broken now for the Next/Previous buttons:
![HMf81J2DQg](https://github.com/woocommerce/woocommerce-blocks/assets/905781/bc396fda-d828-402f-a8cc-4d84ef196dd7)

And also, you can move the Pager up and down, not sure if this is intended:
![lJx2zeGaIs](https://github.com/woocommerce/woocommerce-blocks/assets/905781/ca279ecf-7d06-4c97-ac3f-1d65f798bfdd)

gigitux commented:
It is great that E2E tests caught the regression:

![image](https://github.com/woocommerce/woocommerce-blocks/assets/4463174/3cad8685-0020-434c-aea5-da16cce73d10)

thealexandrelara committed:
Merge branch 'trunk' into fix/10781-pager-block-is-not-correctly-aligned-with-large-image-block

thealexandrelara committed:
Add Large Image and Pager to its own group

thealexandrelara committed:
Fix issue when moving around the thumbnails block

thealexandrelara committed:
Fix e2e tests for Product Gallery block

thealexandrelara commented:
Thank you for reviewing the PR and raising this points.

Initially, I had moved this block to be an inner block of the Product Gallery Large Image block; however, that caused more issues than expected so I decided to take a different approach. 

I put the block inside a group together with the Large Image block, this way we can isolate the behavior for each block. I noticed inner blocks of the Large Image block display over the image, so I think it makes sense to keep the Pager block separated from those blocks.

Finally, I had to change the way that we rearrange the blocks when a different Thumbnails position is selected and fixed the e2e tests that were failing.

danieldudzic reviewed (commented):
Thanks for your work on this. This behaves as expected now.

I have one concern though.

Cause now we are moving into multiple wrapper territory:

<img width="1258" alt="Single_Product_‹_Template_‹_Editor_‹_newproductgallery_—_WordPress" src="https://github.com/woocommerce/woocommerce-blocks/assets/905781/6260ae13-b11c-4f79-8155-98bc85822026">

&nbsp;<br />With the first wrapper, there was no way around it, as it needed to be set dynamically based on the `Thumbnails` position setting, but with the second one, I believe we should be able to utilize the `layout`: https://github.com/woocommerce/woocommerce-blocks/discussions/10326#discussioncomment-6544675

Did you investigate this solution? If possible I'd avoid adding another wrapper.

----------

Update: Actually scratch the above, I don't see a way we can not use a wrapper for these items:

<img width="1017" alt="Single_Product_‹_Template_‹_Editor_‹_newproductgallery_—_WordPress" src="https://github.com/woocommerce/woocommerce-blocks/assets/905781/0d9652a2-1de3-423f-9a1b-b6b6a05a1184">

danieldudzic commented on assets/js/blocks/product-gallery/inner-blocks/product-gallery-pager/block.json:
```diff
@@ -8,6 +8,6 @@
     "keywords": [ "WooCommerce" ],
     "textdomain": "woo-gutenberg-products-block",
-	"ancestor": [ "woocommerce/product-gallery" ],
+	"ancestor": [ "woocommerce/product-gallery-large-image" ],
```
This doesn't seem to be correct, as now the Large Image is Pager's sibling.

<img width="414" alt="Single_Product_‹_Template_‹_Editor_‹_newproductgallery_—_WordPress" src="https://github.com/woocommerce/woocommerce-blocks/assets/905781/e59218bc-e91a-486b-b23e-1e7b1d4c9697">

thealexandrelara replied:
Nice catch! The ancestor should be the Product Gallery block, I just updated this line of code

roykho commented:
I am seeing a console error in editor:

```
Cannot update a component (`wp`) while rendering a different component (`Edit`). To locate the bad setState() call inside `Edit`, follow the stack trace as described in https://reactjs.org/link/setstate-in-render
```

Not related, but not sure why there is the Justification settings in two different parts which is abit confusing.

![file.png](https://github.com/woocommerce/woocommerce-blocks/assets/2132595/913c6062-1301-46a5-aa79-67889fa8291d)

danieldudzic reviewed (approved):
One minor change suggestion, but overall excellent work. The updated tests also pass locally.

The **Pager** is positioned correctly, and also actually benefits from sitting inside of the container, as the user will be able to align it to the `left`, `center` and `right`:

![0l9boAm1aV](https://github.com/woocommerce/woocommerce-blocks/assets/905781/63cb327e-60f8-4ffc-a95d-b706360018e0)

thealexandrelara commented:
> I am seeing a console error in editor:
> 
> ```
> Cannot update a component (`wp`) while rendering a different component (`Edit`). To locate the bad setState() call inside `Edit`, follow the stack trace as described in https://reactjs.org/link/setstate-in-render
> ```
I cannot reproduce this, is this error appearing to you only when the Product Gallery block is inserted on the Editor?
 
> Not related, but not sure why there is the Justification settings in two different parts which is abit confusing.
> 
> ![file.png](https://user-images.githubusercontent.com/2132595/265037655-913c6062-1301-46a5-aa79-67889fa8291d.png)

I think this is a common behavior for the Row and Stack block :/

thealexandrelara committed:
Change ancestor block for the Pager block

roykho commented:
Please see:


https://github.com/woocommerce/woocommerce-blocks/assets/2132595/5e9bc553-e757-4ed5-913c-2ec5aa840c8a

thealexandrelara committed:
Merge branch 'trunk' into fix/10781-pager-block-is-not-correctly-aligned-with-large-image-block

thealexandrelara commented:
> Please see:
> 
>  file.mov

Thank you for sharing the video, I tried multiple times to reproduce this behavior but the issue is not occurring for me :/

@danieldudzic are you experiencing this error on your side?

danieldudzic commented:
> > Please see:
> > file.mov
> 
> Thank you for sharing the video, I tried multiple times to reproduce this behavior but the issue is not occurring for me :/
> 
> @danieldudzic are you experiencing this error on your side?

I also wasn't able to replicate the issue.

gigitux commented:
I'm able to replicate the error reported by @roykho:

![image](https://github.com/woocommerce/woocommerce-blocks/assets/4463174/6908aa0a-6dea-4f83-9c9f-5f050f2222f6)

This error is visible in trunk as well, so I think that we can merge this PR (I didn't review it) if the blocker is the error reported. I will open an issue.

gigitux commented:
Issue created https://github.com/woocommerce/woocommerce-blocks/issues/10836

thealexandrelara commented:
Okay, I'll merge this PR and we can investigate and fix that issue in a separate PR. Thank you!
